### PR TITLE
VQGAN implementation and basic training loop

### PR DESCRIPTION
**Pull request by mahalrs**
This PR includes VQGAN implementation. Majority of architecture related code is adapted from https://github.com/CompVis/taming-transformers.

This PR also includes a VQGAN notebook for experimentation and basic fine-tuning.

### CODE REVIEW

**Comment by mahalrs** [comment]
@ArjunKrishna3367 @Waleedprw22 

You can experiment with VQGAN notebook using visual_news_mini dataset.

**Review by ArjunKrishna3367 — DISMISSED** [review]
Everything worked fine for me, just added a way to save/load model.

**Comment by Waleedprw22** [comment]
Going through the different files, it seems like we have all the necessary components to get the VQGAN up and running. I am wondering though if we will still need the generic vocab json and merge txt file from hugging face for this.

File: src/vqgan/quantize.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+        d = torch.sum(z_flattened ** 2, dim=1, keepdim=True) + \
+            torch.sum(self.embedding.weight**2, dim=1) - 2 * \
+            torch.einsum('bd,dn->bn', z_flattened, rearrange(self.embedding.weight, 'n d -> d n'))
+
+        min_encoding_indices = torch.argmin(d, dim=1)
+        z_q = self.embedding(min_encoding_indices).view(z.shape)
+        perplexity = None
+        min_encodings = None
+
+        # compute loss for embedding
```
Comment (Waleedprw22) [inline src/vqgan/quantize.py:108]: just curious, what is this type of loss called? are there other types we can use as well?

**Comment by mahalrs** [comment]
> Going through the different files, it seems like we have all the necessary components to get the VQGAN up and running. I am wondering though if we will still need the generic vocab json and merge txt file from hugging face for this.

I have to look into this. For the VQGAN, we don't need anything. It's working well.

For BART encoder, we are using a pre-trained encoder, so we don't need a vocab unless we choose to fine-tune it. I think we will just freeze encoder weights.

For the BART decoder, we are training it from scratch so we need a vocab for it. We need to map image tokens to numerical representation. I need to look into this. So, let's discuss this and then we can push all vocab files.

**Comment by mahalrs** [comment]
> > Going through the different files, it seems like we have all the necessary components to get the VQGAN up and running. I am wondering though if we will still need the generic vocab json and merge txt file from hugging face for this.
> 
> I have to look into this. For the VQGAN, we don't need anything. It's working well.
> 
> For BART encoder, we are using a pre-trained encoder, so we don't need a vocab unless we choose to fine-tune it. I think we will just freeze encoder weights.
> 
> For the BART decoder, we are training it from scratch so we need a vocab for it. We need to map image tokens to numerical representation. I need to look into this. So, let's discuss this and then we can push all vocab files.

Also, I'm not sure if my conceptual understanding is correct. So, I will work on it for next few days and will get back soon.

**Review by ArjunKrishna3367 — APPROVED** [review]
Thanks for removing the notebook outputs, forgot to do that in my commit.
